File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        int n = 0;
        ListNode tmp = head;
        while (tmp != null) {
            n++;
            tmp = tmp.next;
        }
        int mid = n / 2;
        int curr = 0;
        ListNode prev = null;
        boolean startReversing = false;
        tmp = head;
        while (tmp != null) {
            if (curr >= mid) {
                ListNode next = tmp.next;
                ListNode tmp2 = tmp;
                if (!startReversing) {
                    startReversing = true;
                    prev.next = null;
                    tmp2.next = null;
                } else {
                    tmp.next = prev;
                }
                prev = tmp2;
                if (next == null) break;
                tmp = next;
            } else {
                prev = tmp;
                tmp = tmp.next;
            }
            curr++;
        }
        assert tmp != null : "Unexpected";
        int ans = Integer.MIN_VALUE;
        ListNode tail = tmp;
        while (head != null && tail != null) {
            ans = Math.max(ans, head.val + tail.val);
            head = head.next;
            tail = tail.next;
        }
        return ans; // TC: O(n), SC: O(1)
    }
}

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        // push all node to vector
        vector<int> nodeList;
        while (head) {
            nodeList.emplace_back(head->val);
            head = head->next;
        }
        int i = 0, j = nodeList.size() - 1, maxTwinSum = 0;
        while (i < j) {
            int sum = nodeList[i] + nodeList[j];
            i++; --j;
            maxTwinSum = max(maxTwinSum, sum);
        }
        return maxTwinSum; // TC: O(n), SC: O(n)
    }
};